lib/devise/models/confirmable.rb

```diff
@@ -209,7 +209,10 @@ def confirmation_required?
         #   confirmation_period_valid?   # will always return true
         #
         def confirmation_period_valid?
-          self.class.allow_unconfirmed_access_for.nil? || (confirmation_sent_at && confirmation_sent_at.utc >= self.class.allow_unconfirmed_access_for.ago)
+          return false if self.class.allow_unconfirmed_access_for == 0.days
+          return true if self.class.allow_unconfirmed_access_for.nil?
```

Comment:
WDYT about moving the `nil` check to the top? I know it will work the same way, but it seems reasonable to return first when the value is `nil`, without the need to check whether is equal to `0.days` or not.